Batteries/Data/Char.lean

```diff
@@ -156,3 +157,74 @@ theorem toUpper_eq_of_isLower {c : Char} (h : c.isLower) : c.toUpper = ofNat (c.
     rw [toUpper_eq_of_isLower hl', toLower_eq_of_isUpper hu, toUpper_eq_of_not_isLower hl,
       toNat_ofNat, if_pos hv, Nat.add_sub_cancel, ofNat_toNat]
   · rw [toLower_eq_of_not_isUpper hu]
+
+/-- Case folding for ASCII characters only.
+
+Alphabetic ASCII characters are mapped to their lowercase form, all other characters are left
+unchanged. This agrees with the Unicode case folding algorithm for ASCII characters.
+
+```
+#eval caseFoldAsciiOnly 'A' == 'a'
+#eval caseFoldAsciiOnly 'a' == 'a'
+#eval caseFoldAsciiOnly 'À' == 'À'
+#eval caseFoldAsciiOnly 'à' == 'à'
+#eval caseFoldAsciiOnly '$' == '$'
+```
+-/
+abbrev caseFoldAsciiOnly := Char.toLower
+
+theorem isAlpha_iff_isUpper_or_isLower {c : Char} : c.isAlpha ↔ c.isUpper ∨ c.isLower := by
+  simp [Char.isAlpha]
+
+theorem toLower_eq_of_isLower {c : Char} : c.isLower → c.toLower = c := by
+  simp only [isLower, ge_iff_le, Bool.and_eq_true, decide_eq_true_eq, toLower, ite_eq_right_iff,
+    and_imp]
+  intro h0 _ _ h3
+  have _ : 97 ≤ c.val.toNat ∧ c.val.toNat ≤ 90 := ⟨h0, h3⟩
+  omega
+
+theorem toLower_eq_of_not_isAlpha {c : Char} : ¬c.isAlpha → c.toLower = c := by
+  simp only [isAlpha, isUpper, ge_iff_le, isLower, Bool.or_eq_true, Bool.and_eq_true,
+    decide_eq_true_eq, not_or, not_and, UInt32.not_le, toLower, ite_eq_right_iff, and_imp]
+  intro h0 _ h2 h3
+  have _ : c.val.toNat ≤ 90 ∧ 90 < c.val.toNat:= ⟨h3, h0 h2⟩
+  omega
```

Comment:
Let's temporarily remove these lemmas since #1333 should cover these.

Reply:
I've removed them in order to keep things moving, but I was not able to close those goals with just `simp` or `apply?`, so at least by this (admittedly very crude) metric they may not necessarily be duplicative.